src/Network/Transport/Chan.hs

```diff
@@ -103,55 +281,110 @@ apiNewMulticastGroup state ourAddress = do
 -- subsequent calls to resolveMulticastGroup will fail. This mimicks the fact
 -- that some multicast messages may still be in transit when the group is
 -- deleted.
-createMulticastGroup :: MVar TransportState -> EndPointAddress -> MulticastAddress -> MVar (Set EndPointAddress) -> MulticastGroup
-createMulticastGroup state ourAddress groupAddress group =
-  MulticastGroup { multicastAddress     = groupAddress
-                 , deleteMulticastGroup = modifyMVar_ state $ return . (multigroups ^: Map.delete groupAddress)
-                 , maxMsgSize           = Nothing
-                 , multicastSend        = \payload -> do
-                                            cs <- (^. channels) <$> readMVar state
-                                            es <- readMVar group
-                                            forM_ (Set.elems es) $ \ep -> do
-                                              let ch = cs ^. at ep "Invalid endpoint"
-                                              writeChan ch (ReceivedMulticast groupAddress payload)
-                 , multicastSubscribe   = modifyMVar_ group $ return . Set.insert ourAddress
-                 , multicastUnsubscribe = modifyMVar_ group $ return . Set.delete ourAddress
-                 , multicastClose       = return ()
-                 }
+createMulticastGroup :: TVar TransportState -> EndPointAddress -> MulticastAddress -> TVar (Set EndPointAddress) -> MulticastGroup
+createMulticastGroup state ourAddress groupAddress group = MulticastGroup
+    { multicastAddress     = groupAddress
+    , deleteMulticastGroup = atomically $
+        whenValidTransportState state $ \vst -> do
+          -- XXX best we can do given current broken API, which needs fixing.
+          let lep = fromJust $ vst ^. localEndPointAt ourAddress
+          modifyTVar' (localEndPointState lep) $ \lepst -> case lepst of
+            LocalEndPointValid lepvst ->
+              LocalEndPointValid $ multigroups ^: Map.delete groupAddress $ lepvst
+            LocalEndPointClosed ->
+              LocalEndPointClosed
+    , maxMsgSize           = Nothing
+    , multicastSend        = \payload -> atomically $
+        withValidTransportState state SendFailed $ \vst -> do
+          es <- readTVar group
+          forM_ (Set.elems es) $ \ep -> do
+            let ch = localEndPointChannel $ fromJust $ vst ^. localEndPointAt ep
+            writeTChan ch (ReceivedMulticast groupAddress payload)
```

Comment:
Mutlicast here is not supported yet, `MVar` based implementation was completely broken. So I've just solve all build errors, without going deeper. We have another Pull Request to disable multicast until it will be fixed.